### PR TITLE
add accessibility buffer

### LINKED ISSUE #169853: Explore plain content editable element for terminal buffer instead of navigation mode
Prototype: https://github.com/xtermjs/xterm.js/pull/4340

### PR DESCRIPTION
**Pull request by meganrogge**
fixes #169853

### CODE REVIEW

File: src/vs/workbench/contrib/terminal/browser/terminalActions.ts [hunk 1]
```diff
@@ -382,6 +382,25 @@ export function registerTerminalActions() {
 			}
 		}
 	});
+	registerAction2(class extends Action2 {
+		constructor() {
+			super({
+				id: TerminalCommandId.FocusAccessibilityElement,
```
Comment (meganrogge) [inline src/vs/workbench/contrib/terminal/browser/terminalActions.ts:388]: enter accessibility mode, focus accessibility view

**Comment by Tyriar** [comment]
I notice mouse wheel events still skip the a11y buffer and go to the main terminal, ideally we would change that as well as when you click on it to not switch back to the main terminal view.

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 2]
```diff
@@ -1095,6 +1096,21 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
...
+			return;
+		}
+		this._accessibilityElement = this.xterm.raw.element.querySelector('.xterm-accessibility-buffer') as HTMLElement || undefined;
+		if (!this._accessibilityElement) {
+			return;
+		}
+		const fontFamily = this._configurationService.getValue(TerminalSettingId.FontFamily) || 'Monospace';
+		const fontSize = this._configurationService.getValue(TerminalSettingId.FontSize);
+		this._accessibilityElement.style.fontFamily = fontFamily ? fontFamily.toString() : '';
+		this._accessibilityElement.style.fontSize = fontSize ? fontSize + 'px' : '';
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:1110]: This part is handled by xterm.js?

https://github.com/xtermjs/xterm.js/blob/41dc93a008dc7bc4044020cfa483210891770b50/src/browser/AccessibilityManager.ts#L346-L349

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 3]
```diff
@@ -1095,6 +1096,21 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
 		}
 	}
 
+	focusAccessibilityElement(): void {
+		if (!this.xterm?.raw.element) {
+			return;
+		}
+		this._accessibilityElement = this.xterm.raw.element.querySelector('.xterm-accessibility-buffer') as HTMLElement || undefined;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:1103]:
```suggestion
		this._accessibilityElement = this.xterm.raw.element.querySelector('.xterm-accessibility-buffer') as HTMLElement | undefined;
```

**Review by Tyriar — CHANGES_REQUESTED** [review]

**Comment by meganrogge** [comment]
> I notice mouse wheel events still skip the a11y buffer and go to the main terminal, ideally we would change that as well as when you click on it to not switch back to the main terminal view.

Oh, I thought that was by design as I thought ppl would be using the keyboard to navigate and mouse to escape

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 4]
```diff
@@ -1095,6 +1096,20 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
...
+	focusAccessibilityBuffer(): void {
+		if (!this.xterm?.raw.element) {
+			return;
+		}
+		this._accessibilityBuffer = this.xterm.raw.element.querySelector('.xterm-accessibility-buffer') as HTMLElement || undefined;
+		if (!this._accessibilityBuffer) {
+			return;
+		}
+		// The viewport is undefined when this is focused, so we cannot get the cell height from that. Instead, estimate using the font.
+		const cellHeight = this._configHelper.getFont((this.xterm.raw as any)._core)?.charHeight || undefined;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:1108]: Can you use `this.xterm.getFont()` instead?

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 5]
```diff
@@ -1095,6 +1096,20 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
...
+		this._accessibilityBuffer.style.lineHeight = cellHeight ? (this._configurationService.getValue(TerminalSettingId.LineHeight) as number) * cellHeight + 'px' : '';
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:1109]: getFont should return the lineHeight as well?

**Review by Tyriar — APPROVED** [review]

**Pull request merged**